crates/stark_poseidon/src/hash.rs

```diff
@@ -0,0 +1,75 @@
+use crate::poseidon::{permute, PoseidonState};
+use stark_curve::{FieldElement, FIELD_ZERO};
+
+/// Hashes a number of messages using the Poseidon hash
+pub fn poseidon_hash(msgs: &[FieldElement]) -> FieldElement {
+    let mut state = [FIELD_ZERO, FIELD_ZERO, FIELD_ZERO];
+    for msg in msgs {
+        state[0] += msg;
+        permute(&mut state);
+    }
+    state[0]
+}
+
+/// The PoseidonHasher can build up a hash by appending to state
+pub struct PoseidonHasher {
+    state: PoseidonState,
+}
+
+impl PoseidonHasher {
+    /// Creates a new PoseidonHasher
+    pub fn new() -> PoseidonHasher {
+        PoseidonHasher {
+            state: [FIELD_ZERO, FIELD_ZERO, FIELD_ZERO],
+        }
+    }
+
+    /// Absorbs message into the hash
+    pub fn write(&mut self, msg: FieldElement) {
+        self.state[0] += msg;
+        permute(&mut self.state);
+    }
+
+    /// Extracts a single output
+    pub fn finish(self) -> FieldElement {
+        self.state[0]
+    }
+}
+
+impl Default for PoseidonHasher {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```

Comment:
Am I correct in saying there are two ways of hashing now? Either using `poseidon_hash` or using the builder `PoseidonHasher`?

Should we not only have a single way maybe?

Also are there any optimisations one can make for the hash over two elements only? As in `H(a,b)`?

Reply:
Yes correct, we could probably settle on using `poseidon_hash()` only. However Poseidon is a natural sponge, so thought it natural to wrap it in one. Since we don't know the use case yet, who knows if we only need the sponge.

Right now there are two ways of doing the same thing, but the thing about the sponge is, we can later add a method `extract(&mut self)` that outputs one value (and invoke permute). Hence we basically have a random number generator, that we can reseed whenever wanted (write), and get randomness whenever wanted (extract or finish).